File: src/lib/document-markdown.ts
function decodeEntities(value: string): string {
    return value
        .replace(/&amp;/gi, '&')
        .replace(/&quot;/gi, '"')
        .replace(/&#39;|&apos;/gi, "'")
        .replace(/&lt;/gi, '<')
        .replace(/&gt;/gi, '>')
        .replace(/&nbsp;/gi, ' ')
        .replace(/&#(\d+);/g, (_, raw: string) => String.fromCodePoint(Number(raw) || 32))
        .replace(/&#x([0-9a-f]+);/gi, (_, raw: string) => String.fromCodePoint(Number.parseInt(raw, 16) || 32))
}

function attrs(tag: string): Record<string, string> {
    const output: Record<string, string> = {}
    const re = /([:\w-]+)\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s>]+))/g
    for (const match of tag.matchAll(re)) {
        output[match[1]!.toLowerCase()] = decodeEntities(match[2] ?? match[3] ?? match[4] ?? '')
    }
    return output
}

function absoluteUrl(raw: string | undefined, base: URL): string | null {
    if (!raw) return null
    const value = decodeEntities(raw).trim().replace(/\\\//g, '/')
    if (!value || /^(?:data|blob|javascript):/i.test(value)) return null
    try {
        const parsed = new URL(value, base)
        return /^https?:$/.test(parsed.protocol) ? parsed.toString() : null
    } catch {
        return null
    }
}

function plainText(value: string): string {
    return decodeEntities(value)
        .replace(/<br\s*\/?\s*>/gi, '\n')
        .replace(/<[^>]+>/g, ' ')
        .replace(/[\t\f\v ]+/g, ' ')
        .replace(/\s*\n\s*/g, '\n')
        .trim()
}

function escapeTableCell(value: string): string {
    return value.replace(/\|/g, '\\|').replace(/\s*\n\s*/g, '<br>')
}

function tableToMarkdown(tableHtml: string): string {
    const rows: string[][] = []
    for (const rowMatch of tableHtml.matchAll(/<tr\b[^>]*>([\s\S]*?)<\/tr>/gi)) {
        const cells: string[] = []
        for (const cellMatch of rowMatch[1]!.matchAll(/<(?:th|td)\b[^>]*>([\s\S]*?)<\/(?:th|td)>/gi)) {
            cells.push(escapeTableCell(plainText(cellMatch[1] || '')))
        }
        if (cells.length) rows.push(cells)
    }
    if (!rows.length) return plainText(tableHtml)
    const width = Math.max(...rows.map((row) => row.length))
    const normalized = rows.map((row) => [...row, ...Array(Math.max(0, width - row.length)).fill('')])
    const lines = [
        `| ${normalized[0]!.join(' | ')} |`,
        `| ${Array(width).fill('---').join(' | ')} |`,
        ...normalized.slice(1).map((row) => `| ${row.join(' | ')} |`),
    ]
    return `\n\n${lines.join('\n')}\n\n`
}

function wechatArticleFragment(html: string): string {
    const marker = html.search(/\bid=["']js_content["']/i)
    if (marker < 0) return ''
    const start = html.lastIndexOf('<', marker)
    if (start < 0) return ''
    const tail = html.slice(start)
    const endCandidates = [
        tail.search(/\bid=["']js_toobar3["']/i),
        tail.search(/\bclass=["'][^"']*rich_media_tool/i),
        tail.search(/<script\b/i),
    ].filter((value) => value > 0)
    const end = endCandidates.length ? Math.min(...endCandidates) : Math.min(tail.length, 2_000_000)
    return tail.slice(0, end)
}

export function htmlFragmentToMarkdown(fragment: string, sourceUrl: string): string {
    if (!fragment.trim()) return ''
    let base: URL
    try {
        base = new URL(sourceUrl)
    } catch {
        return ''
    }

    let value = fragment
        .replace(/<(?:script|style|noscript)\b[^>]*>[\s\S]*?<\/(?:script|style|noscript)>/gi, '')
        .replace(/<table\b[^>]*>[\s\S]*?<\/table>/gi, (table) => tableToMarkdown(table))
        .replace(/<pre\b[^>]*>([\s\S]*?)<\/pre>/gi, (_, body: string) => `\n\n\`\`\`\n${decodeEntities(body).replace(/<[^>]+>/g, '').trim()}\n\`\`\`\n\n`)
        .replace(/<img\b[^>]*>/gi, (tag) => {
            const map = attrs(tag)
            const raw = map['data-src'] || map['data-original'] || map['data-lazy-src'] || map['data-actualsrc'] || map.src
            const url = absoluteUrl(raw, base)
            if (!url) return ''
            const alt = plainText(map.alt || map.title || 'image').replace(/[\[\]]/g, '') || 'image'
            return `\n\n![${alt}](${url})\n\n`
        })
        .replace(/<a\b[^>]*>[\s\S]*?<\/a>/gi, (tag) => {
            const open = tag.match(/^<a\b[^>]*>/i)?.[0] || ''
            const map = attrs(open)
            const text = plainText(tag.replace(/^<a\b[^>]*>/i, '').replace(/<\/a>$/i, '')) || map.title || map.href || 'link'
            const url = absoluteUrl(map.href, base)
            return url ? `[${text.replace(/[\[\]]/g, '')}](${url})` : text
        })
        .replace(/<h1\b[^>]*>([\s\S]*?)<\/h1>/gi, (_, body: string) => `\n\n# ${plainText(body)}\n\n`)
        .replace(/<h2\b[^>]*>([\s\S]*?)<\/h2>/gi, (_, body: string) => `\n\n## ${plainText(body)}\n\n`)
        .replace(/<h3\b[^>]*>([\s\S]*?)<\/h3>/gi, (_, body: string) => `\n\n### ${plainText(body)}\n\n`)
        .replace(/<h[4-6]\b[^>]*>([\s\S]*?)<\/h[4-6]>/gi, (_, body: string) => `\n\n#### ${plainText(body)}\n\n`)
        .replace(/<blockquote\b[^>]*>([\s\S]*?)<\/blockquote>/gi, (_, body: string) => {
            const text = plainText(body)
            return `\n\n${text.split('\n').map((line) => `> ${line}`).join('\n')}\n\n`
        })
        .replace(/<li\b[^>]*>([\s\S]*?)<\/li>/gi, (_, body: string) => `\n- ${plainText(body)}`)
        .replace(/<(?:strong|b)\b[^>]*>([\s\S]*?)<\/(?:strong|b)>/gi, (_, body: string) => `**${plainText(body)}**`)
        .replace(/<(?:em|i)\b[^>]*>([\s\S]*?)<\/(?:em|i)>/gi, (_, body: string) => `*${plainText(body)}*`)
        .replace(/<code\b[^>]*>([\s\S]*?)<\/code>/gi, (_, body: string) => `\`${plainText(body).replace(/`/g, '\\`')}\``)
        .replace(/<br\s*\/?\s*>/gi, '\n')
        .replace(/<\/(?:p|div|section|article|ul|ol)>/gi, '\n\n')
        .replace(/<(?:p|div|section|article|ul|ol)\b[^>]*>/gi, '')
        .replace(/<[^>]+>/g, '')

    value = decodeEntities(value)
        .replace(/\r/g, '')
        .replace(/[\t\f\v ]+\n/g, '\n')
        .replace(/\n[\t\f\v ]+/g, '\n')
        .replace(/\n{3,}/g, '\n\n')
        .trim()

    return value ? `${value}\n` : ''
}

export function extractDocumentMarkdown(sourceUrl: string, html: string, platform: string): string {
    if (platform !== 'wechat') return ''
    return htmlFragmentToMarkdown(wechatArticleFragment(html), sourceUrl)
}

File: local-engine/document_markdown.py
from __future__ import annotations

import html as html_lib
import re
from urllib.parse import urljoin, urlparse


def _attrs(tag: str) -> dict[str, str]:
    result: dict[str, str] = {}
    pattern = re.compile(r"([:\w-]+)\s*=\s*(?:\"([^\"]*)\"|'([^']*)'|([^\s>]+))")
    for match in pattern.finditer(tag):
        result[match.group(1).lower()] = html_lib.unescape(match.group(2) or match.group(3) or match.group(4) or "")
    return result


def _absolute_url(raw: str | None, base_url: str) -> str | None:
    value = html_lib.unescape(str(raw or "")).strip().replace("\\/", "/")
    if not value or re.match(r"^(?:data|blob|javascript):", value, re.I):
        return None
    target = urljoin(base_url, value)
    parsed = urlparse(target)
    return target if parsed.scheme in {"http", "https"} and parsed.hostname else None


def _plain(value: str) -> str:
    text = html_lib.unescape(value or "")
    text = re.sub(r"<br\s*/?>", "\n", text, flags=re.I)
    text = re.sub(r"<[^>]+>", " ", text)
    text = re.sub(r"[\t\f\v ]+", " ", text)
    text = re.sub(r"\s*\n\s*", "\n", text)
    return text.strip()


def _table_markdown(table_html: str) -> str:
    rows: list[list[str]] = []
    for row_match in re.finditer(r"<tr\b[^>]*>([\s\S]*?)</tr>", table_html, re.I):
        cells = [
            _plain(cell.group(1)).replace("|", "\\|").replace("\n", "<br>")
            for cell in re.finditer(r"<(?:th|td)\b[^>]*>([\s\S]*?)</(?:th|td)>", row_match.group(1), re.I)
        ]
        if cells:
            rows.append(cells)
    if not rows:
        return _plain(table_html)
    width = max(len(row) for row in rows)
    normalized = [row + [""] * (width - len(row)) for row in rows]
    lines = [
        "| " + " | ".join(normalized[0]) + " |",
        "| " + " | ".join(["---"] * width) + " |",
        *("| " + " | ".join(row) + " |" for row in normalized[1:]),
    ]
    return "\n\n" + "\n".join(lines) + "\n\n"


def _wechat_fragment(raw_html: str) -> str:
    marker = re.search(r"\bid=[\"']js_content[\"']", raw_html, re.I)
    if marker is None:
        return ""
    start = raw_html.rfind("<", 0, marker.start())
    if start < 0:
        return ""
    tail = raw_html[start:]
    candidates: list[int] = []
    for pattern in (
        r"\bid=[\"']js_toobar3[\"']",
        r"\bclass=[\"'][^\"']*rich_media_tool",
        r"<script\b",
    ):
        match = re.search(pattern, tail, re.I)
        if match and match.start() > 0:
            candidates.append(match.start())
    end = min(candidates) if candidates else min(len(tail), 2_000_000)
    return tail[:end]


def html_fragment_to_markdown(fragment: str, source_url: str) -> str:
    if not fragment.strip():
        return ""
    value = re.sub(r"<(?:script|style|noscript)\b[^>]*>[\s\S]*?</(?:script|style|noscript)>", "", fragment, flags=re.I)
    value = re.sub(r"<table\b[^>]*>[\s\S]*?</table>", lambda match: _table_markdown(match.group(0)), value, flags=re.I)
    value = re.sub(
        r"<pre\b[^>]*>([\s\S]*?)</pre>",
        lambda match: "\n\n```\n" + html_lib.unescape(re.sub(r"<[^>]+>", "", match.group(1))).strip() + "\n```\n\n",
        value,
        flags=re.I,
    )

    def image_replace(match: re.Match[str]) -> str:
        tag = match.group(0)
        values = _attrs(tag)
        raw = values.get("data-src") or values.get("data-original") or values.get("data-lazy-src") or values.get("data-actualsrc") or values.get("src")
        target = _absolute_url(raw, source_url)
        if not target:
            return ""
        alt = _plain(values.get("alt") or values.get("title") or "image").replace("[", "").replace("]", "") or "image"
        return f"\n\n![{alt}]({target})\n\n"

    value = re.sub(r"<img\b[^>]*>", image_replace, value, flags=re.I)

    def link_replace(match: re.Match[str]) -> str:
        tag = match.group(0)
        opening = re.match(r"<a\b[^>]*>", tag, re.I)
        values = _attrs(opening.group(0) if opening else "")
        inner = re.sub(r"^<a\b[^>]*>|</a>$", "", tag, flags=re.I)
        text = _plain(inner) or values.get("title") or values.get("href") or "link"
        target = _absolute_url(values.get("href"), source_url)
        return f"[{text.replace('[', '').replace(']', '')}]({target})" if target else text

    value = re.sub(r"<a\b[^>]*>[\s\S]*?</a>", link_replace, value, flags=re.I)
    for level in range(1, 7):
        prefix = "#" * min(level, 4)
        value = re.sub(
            rf"<h{level}\b[^>]*>([\s\S]*?)</h{level}>",
            lambda match, p=prefix: f"\n\n{p} {_plain(match.group(1))}\n\n",
            value,
            flags=re.I,
        )
    value = re.sub(
        r"<blockquote\b[^>]*>([\s\S]*?)</blockquote>",
        lambda match: "\n\n" + "\n".join(f"> {line}" for line in _plain(match.group(1)).splitlines()) + "\n\n",
        value,
        flags=re.I,
    )
    value = re.sub(r"<li\b[^>]*>([\s\S]*?)</li>", lambda match: "\n- " + _plain(match.group(1)), value, flags=re.I)
    value = re.sub(r"<(?:strong|b)\b[^>]*>([\s\S]*?)</(?:strong|b)>", lambda match: f"**{_plain(match.group(1))}**", value, flags=re.I)
    value = re.sub(r"<(?:em|i)\b[^>]*>([\s\S]*?)</(?:em|i)>", lambda match: f"*{_plain(match.group(1))}*", value, flags=re.I)
    value = re.sub(r"<code\b[^>]*>([\s\S]*?)</code>", lambda match: "`" + _plain(match.group(1)).replace("`", "\\`") + "`", value, flags=re.I)
    value = re.sub(r"<br\s*/?>", "\n", value, flags=re.I)
    value = re.sub(r"</(?:p|div|section|article|ul|ol)>", "\n\n", value, flags=re.I)
    value = re.sub(r"<(?:p|div|section|article|ul|ol)\b[^>]*>", "", value, flags=re.I)
    value = re.sub(r"<[^>]+>", "", value)
    value = html_lib.unescape(value).replace("\r", "")
    value = re.sub(r"[\t\f\v ]+\n", "\n", value)
    value = re.sub(r"\n[\t\f\v ]+", "\n", value)
    value = re.sub(r"\n{3,}", "\n\n", value).strip()
    return value + "\n" if value else ""


def extract_document_markdown(source_url: str, raw_html: str, platform: str) -> str:
    if platform != "wechat":
        return ""
    return html_fragment_to_markdown(_wechat_fragment(raw_html), source_url)
